### PR TITLE
Add support for getting constellations and PrecessedGeocentric frame

### PR DESCRIPTION
**Pull request by eteq**
This PR does two related things to coordinates:

1. Adds a ``get_constellation`` convenience function and `SkyCoord` method. They do exactly what it sounds like: determine the constellation the coordinate is in and returns it.  
2. Adds a `PrecessedGeocentric` frame closely related to `GCRS`, but with added precession of the coordinate system.  This is important because it preserves the observer location as being relative to the Earth while allowing for precession.  This is critical for ``get_constellation`` because constellation boundaries are defined relative to the 1875 equinox, and the existing precess-able frames (FK4 and FK5) are barycentric, so you can't do things like asking what constellation the sun is in.  Adding this frame makes that possible and might be useful for other historical astronomy questions.

cc @adrn @astrofrog @taldcroft

### CODE REVIEW

File: astropy/coordinates/funcs.py [hunk 1]
```diff
@@ -176,8 +176,87 @@ def concatenate(coords):
...
-    from .sky_coordinate import SkyCoord
-
     if getattr(coords, 'isscalar', False) or not isiterable(coords):
         raise TypeError('The argument to concatenate must be iterable')
     return SkyCoord(coords)
+
+
+# global dictionary that caches repeatedly-needed info for get_constellation
+_constellation_data = {}
+def get_constellation(coord, short=False):
```
Comment (taldcroft) [inline astropy/coordinates/funcs.py]: What about a more descriptive arg `short_name`?

File: astropy/coordinates/funcs.py [hunk 2]
```diff
@@ -176,8 +176,87 @@ def concatenate(coords):
...
+    """
+    Determines the constellation(s) of the coordinates this `SkyCoord`
```
Comment (taldcroft) [inline astropy/coordinates/funcs.py]: The coord can also be a low-level coord frame object, not just `SkyCoord`.

File: astropy/coordinates/tests/test_funcs.py [hunk 3]
```diff
@@ -49,3 +51,20 @@ def test_concatenate():
 
     with pytest.raises(TypeError):
         concatenate(1*u.deg)
+
+
+def test_constellations():
+    from .. import ICRS, FK5
+    from ..funcs import get_constellation
+
+    inuma = ICRS(9*u.hour, 65*u.deg)
```
Comment (taldcroft) [inline astropy/coordinates/tests/test_funcs.py]: Might be worth including a specific test for Boötes, it being the unicode oddball.

File: astropy/coordinates/builtin_frames/gcrs.py [hunk 4]
```diff
@@ -69,5 +68,59 @@ class GCRS(BaseCoordinateFrame):
     obsgeoloc = QuantityFrameAttribute(default=0, unit=u.m, shape=(3,))
     obsgeovel = QuantityFrameAttribute(default=0, unit=u.m/u.s, shape=(3,))
 
-#The "self-transform" is defined in icrs_cirs_transformations.py, because in
-#the current implementation it goes through ICRS (like CIRS)
+# The "self-transform" is defined in icrs_cirs_transformations.py, because in
+# the current implementation it goes through ICRS (like CIRS)
+
+
+class PrecessedGeocentric(BaseCoordinateFrame):
```
Comment (taldcroft) [inline astropy/coordinates/builtin_frames/gcrs.py]: I don't see any specific tests for this new frame (besides use in constellation code).

**Comment by taldcroft** [comment]
@eteq - finished review.

I never would have guessed that the formal definition of constellation boundaries is rectilinear like this, but I googled it and now I believe.

**Comment by astrofrog** [comment]
@eteq - I think it would be good to add an option to specify which set of constellations to return (defaulting to Roman). Maybe we will want to integrate ones from other cultures one day?

File: astropy/coordinates/funcs.py (hunk 2 above)
Comment (eteq) [inline astropy/coordinates/funcs.py]: Oops, yeah.  The need for this change (and a similar change in "Returns") also means the bit I suggested on astropy-dev probably isn't worth it for this... Will fix

File: astropy/coordinates/funcs.py (hunk 1 above)
Comment (eteq) [inline astropy/coordinates/funcs.py]: Good idea - will change.

File: astropy/coordinates/tests/test_funcs.py (hunk 3 above)
Comment (eteq) [inline astropy/coordinates/tests/test_funcs.py]: Ah, good point.  Added in version I'll push up shortly.

File: astropy/coordinates/builtin_frames/gcrs.py (hunk 4 above)
Comment (eteq) [inline astropy/coordinates/builtin_frames/gcrs.py]: that's right - I left it as implicit via the constellation code.  But I suppose you're right an explicit one is a good idea, too.

**Comment by eteq** [comment]
Ok, I think this is all set - I've addressed @taldcroft's comments, and added an option for future other constellations as @astrofrog suggested (note that the set here is not "Roman", but rather the IAU's official "modern" constellations - it differs some from Greco-Roman constellations in that there are a bunch more in the south where the classical era didn't have names for the constellations).

As you say, @taldcroft, I was also surprised to find this is how they're laid out... But I suppose it makes sense for 1930 there were no computers to try to figure out how curvy boundaries map to the point you're looking at, so square segments are just easier to look up.

**Pull request merged**